databroker/experimental/apischema_schemas.py

```diff
@@ -0,0 +1,117 @@
+from apischema import ValidationError, validator
+
+from dataclasses import dataclass
+from typing import Dict, Generic, List, Optional, TypeVar
+
+from tiled.structures.core import StructureFamily
+
+# from tiled.structures.array import ArrayStructure, ArrayMacroStructure, BuiltinDtype
+from tiled.structures.dataframe import DataFrameStructure
+from tiled.structures.xarray import DataArrayStructure, DatasetStructure
+from tiled.server.pydantic_array import ArrayStructure
+
+import re
+
+# import pytest
+
+
+StrucT = TypeVar("StrucT")
+
+# Map structure family to the associated
+# structure model. This is used by the validator.
+structure_association = {
+    StructureFamily.array: ArrayStructure,
+    StructureFamily.dataframe: DataFrameStructure,
+    StructureFamily.xarray_data_array: DataArrayStructure,
+    StructureFamily.xarray_dataset: DatasetStructure,
+    # StructureFamily.mode?????
+    # ...
+}
+
+# @dataclass
+# class Box(Generic[T]):
+# content: T
+
+
+@dataclass
+class Document(Generic[StrucT]):
+    id: str
```

Comment:
This will still leave the mongo documents with an `_id` as the primary index. If you want to replace that you can use
```
    uid: Optional[str] = Field(alias="_id")
```

Reply:
In bluesky/databroker, we have done well by having an application that means nothing to MongoDB in addition to the ObjectId `_id` that is under Mongo's control. The application id survives translation into other storage formats (like archival in JSON or msgpack) and avoids the ObjectId-specific semantics.

It does make sense to _index_ the `id` column, though.